### DISCUSSION TITLE
Any idea how to find easily values of (deeply) nested keys, when the keys not always exist?

### DISCUSSION THREAD

**Discussion by halloleo** [kellyjonbrazil/jello#31] [Q&A] [answered]
I have a JSON file like this:

```
{"details": {
  "sections": [
    {"fields": [
      { "x": "don't", "v": "Yes!" },
      { "x": "don't", "w": "don't" }
    ]},
    {"fields": [
      { "x": "don't", "v": "Indeed!" },
    ]}],
  "other-stuff": "don't"}
}
```

I need to extract the values to the `"v"` keys, i.e _"Yes!"_ and _"Indeed!"_ Because the `"v"` key is not always present, I do something like this:

```
cat data.json | jello -c '\
res = []
for s in _.details.sections:
    for f in s.fields:
        try:
            res.append(f.v)
        except KeyError:
            pass
res'
```

This works, but feels pretty cumbersome. Is there a  way to gather the values easier then with my approach? Also it would be nice if I didn't need to know how deep my `"v"` keys are located.

**Comment by kellyjonbrazil** [comment, marked as answer]
Your solution works, but I agree it might be nicer to do it without a `try`/`except` block. One thing you can do when you don't know if a key exists is to use the `in` operator on the dictionary. It would look something like this:
```bash
cat data.json | jello -c '\
res = []
for s in _.details.sections:
    for f in s.fields:
        if "v" in f:
            res.append(f.v)
res'
```
The output should be:
```json
["Yes!","Indeed!"]
```
To find the keys no matter how deep they are, you may need to use a recursive method. Fortunately, the `-s` option in `jello` happens to be a recursive function, so you can sort of hack a solution using `-s` with `grep`:
```bash
$ cat data.json | jello -s | grep ".v ="
.details.sections[0].fields[0].v = "Yes!";
.details.sections[1].fields[0].v = "Indeed!";
```
Then you would use normal text utilities (`sed`/`awk`/etc.) to grab the data. Or you could use that notation back in `jello` to get those values back into a list:
```bash
$ cat data.json | jello -c '\
res = []
res.append(_.details.sections[0].fields[0].v)
res.append(_.details.sections[1].fields[0].v)
res'
```
```json
["Yes!","Indeed!"]
```

Happy JSON filtering!

**Reply by halloleo** [reply]
Thanks for your reply, @kellyjonbrazil. I like in particular your idea to utilise `-s`!

And I agree, this question is better placed here in the Discussions. Sorry!

**Reply by kellyjonbrazil** [reply]
No worries - any time! :)

**Reply by halloleo** [reply]
BTW, just wanted to mention: I really like the concept of jello!

For a while I have used `jq` on the command line but felt always unfamiliar with it; then I have used plain Python for mangling JSON, but getting the data from and to the command line is slightly cumbersome, particular because piping is difficult.  -  jello combines the two worlds beautifully!
